File: backend/src/worker/scheduler.rs
use std::sync::Arc;
use tokio::time::{interval, Duration};
use crate::{app::AppState, db::queries, api::events::SseEvent, pipeline::trigger};
use crate::notifications::{self, telegram::TelegramEvent};

const ACTIVE_STATUSES: &[&str] = &[
    "processing", "downloading", "checkpoint_download",
    "transcoding", "checkpoint_transcode", "uploading",
];

pub async fn scheduler_loop(state: Arc<AppState>) {
    let mut ticker = interval(Duration::from_secs(15));
    loop {
        ticker.tick().await;
        if let Err(e) = scheduler_tick(state.clone()).await {
            tracing::error!("Scheduler tick error: {}", e);
        }
    }
}

async fn scheduler_tick(state: Arc<AppState>) -> Result<(), Box<dyn std::error::Error>> {
    let channel_count = get_channel_count(&state).await;
    let max_concurrent = std::cmp::max(1, channel_count);
    let active_count = queries::count_jobs_by_statuses(&state.db, ACTIVE_STATUSES).await?;
    let slots = max_concurrent.saturating_sub(active_count as usize);

    if slots == 0 {
        broadcast_queue_update(&state).await?;
        return Ok(());
    }

    tracing::info!("Active: {}, slots remaining: {}", active_count, slots);

    for _ in 0..slots {
        let job = match queries::get_next_queued_job(&state.db).await? {
            Some(j) => j,
            None => break,
        };

        queries::update_job_status(&state.db, &job.id, "processing").await?;
        queries::insert_job_event(
            &state.db, &job.id, None, "status_change",
            "Pipeline started by scheduler", None,
        ).await?;

        let _ = state.event_tx.send(SseEvent::JobStarted {
            job_id: job.id.clone(),
        });

        let title = job.title.clone().unwrap_or_default();
        notifications::send_notification(&state, TelegramEvent::JobStarted(title));

        match trigger::trigger_pipeline(&state, &job, false, false).await {
            Ok(run_id) => {
                tracing::info!("Triggered GHA run {} for job {}", run_id, job.id);
            }
            Err(e) => {
                tracing::error!("Failed to trigger pipeline for job {}: {}", job.id, e);
                queries::update_job_failed(&state.db, &job.id, &format!("Trigger failed: {}", e)).await?;
            }
        }
    }

    broadcast_queue_update(&state).await?;
    Ok(())
}

/// Count configured Discord channels.
/// `discord_channel_ids` → comma-count, fallback `discord_channel_id` → 1, else 1.
/// ponytail: global cap, per-account caps if multi-user needed later
async fn get_channel_count(state: &Arc<AppState>) -> usize {
    if let Ok(Some(ids)) = trigger::get_setting_or_env(state, "discord_channel_ids").await {
        let count = ids.split(',').filter(|c| !c.trim().is_empty()).count();
        if count > 0 { return count; }
    }
    if let Ok(Some(id)) = trigger::get_setting_or_env(state, "discord_channel_id").await {
        if !id.is_empty() { return 1; }
    }
    1
}

async fn broadcast_queue_update(state: &Arc<AppState>) -> Result<(), Box<dyn std::error::Error>> {
    let processing_count = queries::count_jobs_by_status(&state.db, "processing").await?;
    let queued_count = queries::count_jobs_by_status(&state.db, "queued").await?;
    let _ = state.event_tx.send(SseEvent::QueueUpdate {
        processing: processing_count as usize,
        queued: queued_count as usize,
    });
    Ok(())
}

File: backend-bun/src/worker/scheduler.ts
import type { DrizzleDB } from "../db/index";
import type { EventBus } from "../api/events";
import type { Config } from "../config";
import type { Context } from "hono";
import type { AppBindings } from "../app";
import {
  countJobsByStatuses,
  countJobsByStatus,
  getNextQueuedJob,
  updateJobStatus,
  insertJobEvent,
  updateJobFailed,
} from "../db/queries";
import { triggerPipeline } from "../pipeline/trigger";
import { sendNotification } from "../notifications/telegram";

const ACTIVE_STATUSES = [
  "processing",
  "downloading",
  "checkpoint_download",
  "transcoding",
  "checkpoint_transcode",
  "uploading",
];

const TICK_MS = 15_000;

function buildFakeContext(
  db: DrizzleDB,
  config: Config,
  eventBus: EventBus,
): Context<AppBindings> {
  // triggerPipeline only reads c.var.{db,config} via getSettingOrEnv / getDiscordChannel
  // Hono's Context is a class we can't construct outside the framework;
  // this minimal shape satisfies all internal access patterns.
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
  return {
    var: { db, config, eventBus },
  } as unknown as Context<AppBindings>;
}

function broadcastQueueUpdate(db: DrizzleDB, eventBus: EventBus): void {
  const processing = countJobsByStatuses(db, ACTIVE_STATUSES);
  const queued = countJobsByStatus(db, "queued");
  eventBus.send({ type: "queue_update", data: { processing, queued } });
}

function computeMaxConcurrent(config: Config): number {
  let channelCount = 0;
  if (config.discordChannelIds) {
    channelCount = Math.max(
      0,
      config.discordChannelIds.split(",").length,
    );
  } else if (config.discordChannelId) {
    channelCount = 1;
  } else {
    channelCount = 1;
  }
  return Math.max(1, channelCount);
}

async function tick(
  db: DrizzleDB,
  config: Config,
  eventBus: EventBus,
): Promise<void> {
  const maxConcurrent = computeMaxConcurrent(config);
  const activeCount = countJobsByStatuses(db, ACTIVE_STATUSES);
  const slots = maxConcurrent - activeCount;

  if (slots <= 0) {
    broadcastQueueUpdate(db, eventBus);
    return;
  }

  for (let i = 0; i < slots; i++) {
    const job = getNextQueuedJob(db);
    if (!job) break;

    updateJobStatus(db, job.id, "processing");
    insertJobEvent(
      db,
      job.id,
      null,
      "status_change",
      "Pipeline started by scheduler",
      null,
    );
    eventBus.send({ type: "job_started", data: { job_id: job.id } });
    // ponytail: sendNotification is async; fire-and-forget is fine here
    sendNotification(null as unknown as Context<AppBindings>, {
      type: "JobStarted",
      title: job.title ?? "",
    }).catch(() => {});

    const ctx = buildFakeContext(db, config, eventBus);
    triggerPipeline(ctx, job, false, false).catch((e: unknown) => {
      const msg = e instanceof Error ? e.message : String(e);
      updateJobFailed(db, job.id, `Trigger failed: ${msg}`);
    });
  }

  broadcastQueueUpdate(db, eventBus);
}

/**
 * Start the scheduler loop. Returns a cleanup function to stop it.
 */
export function worker(
  db: DrizzleDB,
  config: Config,
  eventBus: EventBus,
): () => void {
  const timer = setInterval(() => {
    tick(db, config, eventBus).catch((e) => {
      console.error("[scheduler] tick error:", e);
    });
  }, TICK_MS);
  return () => clearInterval(timer);
}
